cranelift/codegen/src/ir/pcc.rs

```diff
@@ -0,0 +1,347 @@
+            }),
+            _ => None,
+        }
+    }
+
+    /// Computes the `sextend` of a value with the given facts.
+    pub fn sextend(&self, from_width: u16, to_width: u16) -> Option<Self> {
+        match self {
+            // If we have a defined value in bits 0..bit_width, and
+            // the MSB w.r.t. `from_width` is *not* set, then we can
+            // do the same as `uextend`.
+            Fact::ValueMax { bit_width, max }
+                if *bit_width == from_width && (*max & (1 << (*bit_width - 1)) == 0) =>
+            {
+                self.uextend(from_width, to_width)
+            }
+            _ => None,
+        }
+    }
+
+    /// Scales a value with a fact by a known constant.
+    pub fn scale(&self, width: u16, factor: u32) -> Option<Self> {
+        match self {
+            Fact::ValueMax { bit_width, max } if *bit_width == width => {
+                let max = max.checked_mul(factor as u64)?;
+                if *bit_width < 64 && max > ((1 << width) - 1) {
+                    return None;
+                }
+                Some(Fact::ValueMax {
+                    bit_width: *bit_width,
+                    max,
+                })
+            }
+            _ => None,
+        }
+    }
+
+    /// Offsets a value with a fact by a known amount.
+    pub fn offset(&self, width: u16, offset: i64) -> Option<Self> {
+        // If we eventually support two-sided ranges, we can
+        // represent (0..n) + m -> ((0+m)..(n+m)). However,
+        // right now, all ranges start with zero, so any
+        // negative offset could underflow, and removes all
+        // claims of constrained range.
+        if offset < 0 {
+            return None;
+        }
+        let offset = u64::try_from(offset).unwrap();
+
+        match self {
+            Fact::ValueMax { bit_width, max } if *bit_width == width => {
+                let max = max.checked_add(offset).unwrap();
+                Some(Fact::ValueMax {
+                    bit_width: *bit_width,
+                    max,
+                })
+            }
+            Fact::PointsTo {
+                region: MemoryRegion { max },
+            } => {
+                let max = max.checked_sub(offset).unwrap();
+                Some(Fact::PointsTo {
+                    region: MemoryRegion { max },
+                })
+            }
+            _ => None,
+        }
+    }
+
+    /// Check that accessing memory via a pointer with this fact, with
+    /// a memory access of the given size, is valid.
+    pub fn check_address(&self, size: u32) -> PccResult<()> {
+        match self {
+            Fact::PointsTo {
+                region: MemoryRegion { max },
+            } => ensure!(u64::from(size) <= *max, OutOfBounds),
+            _ => bail!(OutOfBounds),
+        }
+
+        Ok(())
+    }
+}
+
+/// Top-level entry point after compilation: this checks the facts in
+/// VCode.
+pub fn check_facts<B: LowerBackend + TargetIsa>(
```

Comment:
Maybe name this `check_vcode_facts` to leave the door open for `check_clif_facts` as well?

Reply:
Done!